File: problem_sets/ps1/ps1a.clj
(ns ps1a
  (:require
   [clojure.math.combinatorics :as combo]
   [clojure.pprint]
   [clojure.string :as s]))

;;; Read in cow data file
(defn load-cows
  "Read the cow file into a map."
  [cow-file]
  (->>
   cow-file
   slurp
   s/split-lines
   (map #(s/split % #",")) ; [cow weight] lists
   ;; Make the weight an integer and roll the pairs into a hashmap
   (reduce #(assoc %1 (first %2) (Integer/parseInt (second %2))) {})))

;;; First, the greedy algorithm


;;; Single knapsack filler
(defn greedy-fill-knapsack
  "Greedily add one item to knapsack.
  knapsack is a vector of items in the bag.
  remaining-weight is the weight the bag can still accept
  items are the remaining items outside the bag.
  "
  [knapsack remaining-weight items]
  ;; light-items are light enough to fit in the bag
  (let [[light-items heavy-items]
        (map vec (split-with #(<= (second %) remaining-weight) items))]
    (if (empty? light-items)
      ;; No items light enough to fit in the bag,
      ;; return knapsack and remaining items
      [knapsack items]
      ;; Add the heaviest item that'll fit into the bag.
      ;; Decrement the remaining weight, remove item
      ;; from remaining items.
      (greedy-fill-knapsack
       (conj knapsack (peek light-items))
       (- remaining-weight (second (peek light-items)))
       (into (pop light-items) heavy-items)))))

;;; Greedy knapsack problem
(defn greedy-cow-transport
  "Greedily fill all knapsacks.
  knapsacks is a list of knapsacks filled so far
  items: remaining items and weights, sorted by weight
  limit: weight limit"
  ([knapsacks items limit]
  (if
    (empty? items)
    knapsacks
    ;; With items remaining, fill another knapsack.
    (let [[knapsack remaining-items]
          (greedy-fill-knapsack [] limit items)]
      (greedy-cow-transport
       (conj knapsacks knapsack) remaining-items limit))))
  ;; Initial call, start with empty list of knapsacks
  ([items limit]
   (greedy-cow-transport [] (sort-by val items) limit))
  ([items]
   (greedy-cow-transport items 10)))


;;; Now the brute force algorithm

(defn barge-weight
  "Get the total weight of a barge"
  [barge]
  (apply + (map second barge)))

(defn valid-barges?
  "True if no barges are over the limit."
  [limit barges]
  (every? (partial >= limit) (map barge-weight barges)))

;;; Brute force knapsack problem
(defn brute-force-cow-transport
  "Find the optimal transport situation by brute force.
  Starting with the fewest number of barges and going up,
  test every possible transport situation until one works.
  cows: dictionary of cow-name: weight
  limit: max weight per barge."
  ([cows limit]
   (first
    (filter (partial valid-barges? limit)
            (combo/partitions cows)))
   )
  ([cows] (brute-force-cow-transport cows 10))
  )


;;; Comparing algorithms
(defn compare-cow-transport-algorithms
  "Print out a comparision of the time it took the algos to run."
  []
  
  (let [algos [greedy-cow-transport brute-force-cow-transport]
        cow-list (load-cows "ps1/ps1_cow_data.txt")]
    (doseq [algo algos]
      (print (str algo " running. "))
      (time (algo cow-list)))))

(defn test-algos
  []
  (def cow-list (load-cows "ps1/ps1_cow_data.txt"))
  (println cow-list)
  ;;(def el-list {"a" 23 "b" 34 "c" 36 "d" 45 "e" 56 "f" 68})
  ;;(println (combo/partitions [1 2 3]))
  (print "Greedy: ")
  (clojure.pprint/pprint (greedy-cow-transport cow-list 10))
  (print "Brute-force: ")
  (clojure.pprint/pprint (brute-force-cow-transport cow-list))
  (compare-cow-transport-algorithms))

 (test-algos)
     
     

File: problem_sets/ps1/ps1a.py
###########################
# 6.0002 Problem Set 1a: Space Cows 
# Name:
# Collaborators:
# Time:

from ps1_partition import get_partitions
import time
import functools

#================================
# Part A: Transporting Space Cows
#================================

# Separate out the cows from their weight in one line
def cow_line_splitter(cow_line):
    """
    Turn a single line from the cow data file into a list of [cow, weight].
    """
    return cow_line.strip().split(",")
    

# Problem 1
def load_cows(filename):
    """
    Read the contents of the given file.  Assumes the file contents contain
    data in the form of comma-separated cow name, weight pairs, and return a
    dictionary containing cow names as keys and corresponding weights as values.

    Parameters:
    filename - the name of the data file as a string

    Returns:
    a dictionary of cow name (string), weight (int) pairs
    """
 
    with open(filename) as cow_file:
        cow_lines = cow_file.readlines()
        # List of [cow, weight] datapoints.
        cows = [cow_line_splitter(cow_line) for cow_line in cow_lines]
        # Dictionary of {cow: weight} pairs.
        cow_dict = {cow[0]:int(cow[1]) for cow in cows}
    return cow_dict

# Problem 2
def greedy_cow_transport(cows,limit=10):
    """
    Uses a greedy heuristic to determine an allocation of cows that attempts to
    minimize the number of spaceship trips needed to transport all the cows. The
    returned allocation of cows may or may not be optimal.
    The greedy heuristic should follow the following method:

    1. As long as the current trip can fit another cow, add the largest cow that will fit
        to the trip
    2. Once the trip is full, begin a new trip to transport the remaining cows

    Does not mutate the given dictionary of cows.

    Parameters:
    cows - a dictionary of name (string), weight (int) pairs
    limit - weight limit of the spaceship (an int)
    
    Returns:
    A list of lists, with each inner list containing the names of cows
    transported on a particular trip and the overall list containing all the
    trips
    """

    # Get new sorted list; ensure it's still a list of lists rather than tuples.
    sorted_cows = list(map(list, sorted(cows.items(), key=lambda item: item[1], reverse=True)))
    transports = []
    while sorted_cows:
        # New barge
        transport = []
        remaining_weight = limit
        while remaining_weight > 0:
            # Get the heaviest cow that'll fit on the barge.
            next_cow = next(filter(lambda cow: cow[1] <= remaining_weight, sorted_cows), None)
            
            if next_cow:
                # There is a cow that fits, so add it to the transport and
                # remove it from the remaining cows
                transport.append(next_cow)
                sorted_cows.remove(next_cow)
                remaining_weight -= next_cow[1]
            else:
                # No cow fits, so move onto the next transport
                break
        # Add the full transport to the list of transports
        transports.append(transport)
            
    return transports

# Problem 3
def brute_force_cow_transport(cows,limit=10):
    """
    Finds the allocation of cows that minimizes the number of spaceship trips
    via brute force.  The brute force algorithm should follow the following method:

    1. Enumerate all possible ways that the cows can be divided into separate trips 
        Use the given get_partitions function in ps1_partition.py to help you!
    2. Select the allocation that minimizes the number of trips without making any trip
        that does not obey the weight limitation
            
    Does not mutate the given dictionary of cows.

    Parameters:
    cows - a dictionary of name (string), weight (int) pairs
    limit - weight limit of the spaceship (an int)
    
    Returns:
    A list of lists, with each inner list containing the names of cows
    transported on a particular trip and the overall list containing all the
    trips
    """
    # Empty dictionary, return an empty list
    if not(cows): return []
    # Convert the dictionary into a list of tuples (tuples are hashable, and
    # get_partitions uses sets).
    cow_list = [(k,v) for k,v in cows.items()]
    # Generator of all possible partitions
    cow_partitions = get_partitions(cow_list)
    for partition in cow_partitions:
        # Create a list of transport weights for this partition
        barge_weights = [sum([cow[1] for cow in barge]) for barge in partition]
        # If all transports are within the limit, we're done.
        if max(barge_weights) <= limit:
            return partition

    # Some cow must be over the limit.
    return None
        
# Problem 4
def compare_cow_transport_algorithms():
    """
    Using the data from ps1_cow_data.txt and the specified weight limit, run your
    greedy_cow_transport and brute_force_cow_transport functions here. Use the
    default weight limits of 10 for both greedy_cow_transport and
    brute_force_cow_transport.
    
    Print out the number of trips returned by each method, and how long each
    method takes to run in seconds.

    Returns:
    Does not return anything.
    """
    algos = [greedy_cow_transport, brute_force_cow_transport]
    cows = load_cows("/ps1_cow_data.txt")
    for algo in algos:
        start = time.time()
        algo(cows)
        stop = time.time()
        print(algo.__name__, "ran in", 1000*(stop-start), "ms.")
    pass
